File: 108-convert-sorted-array-to-binary-search-tree/108-convert-sorted-array-to-binary-search-tree.java

class Solution {
    public TreeNode sortedArrayToBST(int[] num) {
            if (num.length == 0) {
          return null;
      }
      TreeNode root = arrayToTreeConverter(num, 0, num.length - 1);
      return root;
    }
    public TreeNode arrayToTreeConverter(int[] num, int low, int high) {
        if (low > high) { 
            return null;
        }
        int mid = (low + high) / 2;
        TreeNode root = new TreeNode(num[mid]);
        root.left = arrayToTreeConverter(num, low, mid - 1);
        root.right = arrayToTreeConverter(num, mid + 1, high);
        return root;
    }
}

File: GO-Leet/108. Convert Sorted Array to Binary Search Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
package main

func sortedArrayToBST(nums []int) *TreeNode {
	var create func(int, int) *TreeNode
	create = func(start int, end int) *TreeNode {
		if start > end {
			return nil
		}

		mid := (start + end) / 2

		root := &TreeNode{Val: nums[mid]}

		root.Left = create(start, mid-1)
		root.Right = create(mid+1, end)

		return root
	}

	return create(0, len(nums)-1)
}
